Clean up RemoteCache classes

aorenste opened the pull request:
Summary:
The existing RemoteCacheBackend classes were a bit haphazard - some of them accepted bytes only, some accepted objects, some returned different types of objects than were passed in.

Update them to be more consistent:

1. RemoteCacheBackend is an implementation of a backend: Redis, Memcache, Manifold, LocalFile

2. RemoteCacheSerde is an implementation of a serde protocol - to turn structured objects (dict, list, etc) into bytes: RemoteCacheJsonSerde (json encoding), RemoteCachePassthroughSerde (strictly bytes only)

3. RemoteCache is the cache implementation itself, mixing a RemoteCacheBackend along with an RemoteCacheSerde to provide structured caching.

Other than simply reorganizing the existing cache code this also fixes the Redis autotune caching for OSS.

Test Plan: unit tests

Reviewed By: oulgen

Differential Revision: D61178859


cc @voznesenskym @penguinwu @EikanWang @jgong5 @Guobing-Chen @XiaobingSuper @zhuhaozhe @blzheng @wenzhe-nrv @jiayisunx @ipiszy @yf225 @chenyang78 @kadeng @muchulee8 @ColinPeppler @amjames @desertfire @chauhang

s-deeper commented:
This was suggest for `2.4.1` cherry-picking. Can you evaluated the relationship with the crash at https://github.com/pytorch/pytorch/issues/132400#issuecomment-2313878009 ?

aorenste commented:
> This was suggest for `2.4.1` cherry-picking. Can you evaluate the relationship with the crash at [#132400 (comment)](https://github.com/pytorch/pytorch/issues/132400#issuecomment-2313878009) ?

That does seem like a problem fixed by this PR - but the PR isn't ready to land yet (hoping today or tomorrow) so definitely don't cherry-pick it until it lands.

evkogs commented:
> Remote cache is crashing I don't know if we could cherry-pick #134032

Yep, it does, I confirm the issue


```
[rank14]: Traceback (most recent call last):                                                                                                                                                                                                                                                                                                                                              
[rank14]:   File "/home/ubuntu/efs_gpu/libs/sd-scripts/./sdxl_train.py", line 960, in <module>                                                                                                                                                                                                                                                                                            
[rank14]:     train(args)                                                                                                                                                                                                                                                                                                                                                                 
[rank14]:   File "/home/ubuntu/efs_gpu/libs/sd-scripts/./sdxl_train.py", line 747, in train                                                                                                                                                                                                                                                                                               
[rank14]:     accelerator.backward(loss)                                                                                                                                                                                                                                                                                                                                                  
[rank14]:   File "/home/ubuntu/efs_gpu/libs/accelerate/src/accelerate/accelerator.py", line 2195, in backward                                                                                                                                                                                                                                                                             
[rank14]:     self.scaler.scale(loss).backward(**kwargs)                                                                                                                                                                                                                                                                                                                                  
[rank14]:   File "/home/ubuntu/efs_gpu/miniconda3/envs/diffusion_torch2.4/lib/python3.11/site-packages/torch/_tensor.py", line 521, in backward                                                                                                                                                                                                                                           
[rank14]:     torch.autograd.backward(                                                                                                                                                                                                                                                                                                                                                    
[rank14]:   File "/home/ubuntu/efs_gpu/miniconda3/envs/diffusion_torch2.4/lib/python3.11/site-packages/torch/autograd/__init__.py", line 289, in backward                                                                                                                                                                                                                                 
[rank14]:     _engine_run_backward(                                                                                                                                                                                                                                                                                                                                                       
[rank14]:   File "/home/ubuntu/efs_gpu/miniconda3/envs/diffusion_torch2.4/lib/python3.11/site-packages/torch/autograd/graph.py", line 768, in _engine_run_backward                                                                                                                                                                                                                        
[rank14]:     return Variable._execution_engine.run_backward(  # Calls into the C++ engine to run the backward pass                                                                                                                                                                                                                                                                       
[rank14]:            ^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^                                                                                                                                                                                                                                                                       
[rank14]:   File "/home/ubuntu/efs_gpu/miniconda3/envs/diffusion_torch2.4/lib/python3.11/site-packages/torch/autograd/function.py", line 306, in apply                                                                                                                                                                                                                                    
[rank14]:     return user_fn(self, *args)                                                                                                                                                                                                                                                                                                                                                 
[rank14]:            ^^^^^^^^^^^^^^^^^^^^                                                                                                                                                                                                                                                                                                                                                 
[rank14]:   File "/home/ubuntu/efs_gpu/miniconda3/envs/diffusion_torch2.4/lib/python3.11/site-packages/torch/_functorch/_aot_autograd/runtime_wrappers.py", line 1861, in backward                                                                                                                                                                                                        
[rank14]:     out = call_compiled_backward()                                                                                                                                                                                                                                                                                                                                              
[rank14]:           ^^^^^^^^^^^^^^^^^^^^^^^^                                                                                                                                                                                                                                                                                                                                              
[rank14]:   File "/home/ubuntu/efs_gpu/miniconda3/envs/diffusion_torch2.4/lib/python3.11/site-packages/torch/_functorch/_aot_autograd/runtime_wrappers.py", line 1809, in call_compiled_backward                                                                                                                                                                                          
[rank14]:     out = call_func_at_runtime_with_args(                                                                                                                                                                                                                                                                                                                                       
[rank14]:           ^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^                                                                                                                                                                                                                                                                                                                                       
[rank14]:   File "/home/ubuntu/efs_gpu/miniconda3/envs/diffusion_torch2.4/lib/python3.11/site-packages/torch/_functorch/_aot_autograd/utils.py", line 120, in call_func_at_runtime_with_args                                                                                                                                                                                              
[rank14]:     out = normalize_as_list(f(args))                                                                                                                                                                                                                                                                                                                                            
[rank14]:                             ^^^^^^^                                                                                                                                                                                                                                                                                                                                             
[rank14]:   File "/home/ubuntu/efs_gpu/miniconda3/envs/diffusion_torch2.4/lib/python3.11/site-packages/torch/_dynamo/eval_frame.py", line 600, in _fn                                                                                                                                                                                                                                     
[rank14]:     return fn(*args, **kwargs)                                                                                                                                                                                                                                                                                                                                                  
[rank14]:            ^^^^^^^^^^^^^^^^^^^                                                                                                                                                                                                                                                                                                                                                  
[rank14]:   File "/home/ubuntu/efs_gpu/miniconda3/envs/diffusion_torch2.4/lib/python3.11/site-packages/torch/_inductor/codecache.py", line 1131, in __call__                                                                                                                                                                                                                              
[rank14]:     return self.current_callable(inputs)                                                                                                                                                                                                                                                                                                                                        
[rank14]:            ^^^^^^^^^^^^^^^^^^^^^^^^^^^^^                                                                                                                                                                                                                                                                                                                                        
[rank14]:   File "/home/ubuntu/efs_gpu/miniconda3/envs/diffusion_torch2.4/lib/python3.11/site-packages/torch/_inductor/compile_fx.py", line 944, in run                                                                                                                                                                                                                                   
[rank14]:     return model(new_inputs)                                                                                                                                                                                                                                                                                                                                                    
[rank14]:            ^^^^^^^^^^^^^^^^^                                                                                                                                                                                                                                                                                                                                                    
[rank14]:   File "/tmp/inductor_cache/7u/c7uxyl6sa2vhqznxqbjjnq6jlo2l4wwfegkn2ofv4b6b5srjtqac.py", line 9192, in call                                                                                                                                                                                                                                                                     
[rank14]:     triton_poi_fused__to_copy_0.run(tangents_1, buf0, 64512, grid=grid(64512), stream=stream2)                                                                                                                                                                                                                                                                                  
[rank14]:   File "/home/ubuntu/efs_gpu/miniconda3/envs/diffusion_torch2.4/lib/python3.11/site-packages/torch/_inductor/runtime/triton_heuristics.py", line 820, in run                                                                                                                                                                                                                    
[rank14]:     self.autotune_to_one_config(*args, grid=grid, **kwargs)                                                                                                                                                                                                                                                                                                                     
[rank14]:   File "/home/ubuntu/efs_gpu/miniconda3/envs/diffusion_torch2.4/lib/python3.11/site-packages/torch/_inductor/runtime/triton_heuristics.py", line 722, in autotune_to_one_config                                                                                                                                                                                                 
[rank14]:     self.save_cache_hook(self.launchers[0].config, time_taken_ns)                                                                                                                                                                                                                                                                                                               
[rank14]:   File "/home/ubuntu/efs_gpu/miniconda3/envs/diffusion_torch2.4/lib/python3.11/site-packages/torch/_inductor/runtime/triton_heuristics.py", line 1122, in save_cache_hook                                                                                                                                                                                                       
[rank14]:     remote_cache.put(remote_cache_key, data)                                                                                                                                                                                                                                                                                                                                    
[rank14]:   File "/home/ubuntu/efs_gpu/miniconda3/envs/diffusion_torch2.4/lib/python3.11/site-packages/torch/_inductor/remote_cache.py", line 44, in put                                                                                                                                                                                                                                  
[rank14]:     return self._redis.set(self._get_key(key), data)                                                                                                                                                                                                                                                                                                                            
[rank14]:            ^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^                                                                                                                                                                                                                                                                                                                            
[rank14]:   File "/home/ubuntu/efs_gpu/miniconda3/envs/diffusion_torch2.4/lib/python3.11/site-packages/redis/commands/core.py", line 2333, in set                                                                                                                                                                                                                                         
[rank14]:     return self.execute_command("SET", *pieces, **options)                                                                                                                                                                                                                                                                                                                      
[rank14]:            ^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^                                                                                                                                                                                                                                                                                                                      
[rank14]:   File "/home/ubuntu/efs_gpu/miniconda3/envs/diffusion_torch2.4/lib/python3.11/site-packages/redis/client.py", line 548, in execute_command                                                                                                                                                                                                                                     
[rank14]:     return conn.retry.call_with_retry(                                                                                                                                                                                                                                                                                                                                          
[rank14]:            ^^^^^^^^^^^^^^^^^^^^^^^^^^^                                                                                                                                                                                                                                                                                                                                          
[rank14]:   File "/home/ubuntu/efs_gpu/miniconda3/envs/diffusion_torch2.4/lib/python3.11/site-packages/redis/retry.py", line 62, in call_with_retry                                                                                                                                                                                                                                       
[rank14]:     return do()                                                                                                                                                                                                                                                                                                                                                                 
[rank14]:            ^^^^                                                                                                                                                                                                                                                                                                                                                                 
[rank14]:   File "/home/ubuntu/efs_gpu/miniconda3/envs/diffusion_torch2.4/lib/python3.11/site-packages/redis/client.py", line 549, in <lambda>                                                                                                                                                                                                                                            
[rank14]:     lambda: self._send_command_parse_response(                                                                                                                                                                                                                                                                                                                                  
[rank14]:             ^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^                                                                                                                                                                                                                                                                                                                                  
[rank14]:   File "/home/ubuntu/efs_gpu/miniconda3/envs/diffusion_torch2.4/lib/python3.11/site-packages/redis/client.py", line 524, in _send_command_parse_response                                                                                                                                                                                                                        
[rank14]:     conn.send_command(*args)                                                                                                                                                                                                                                                                                                                                                    
[rank14]:   File "/home/ubuntu/efs_gpu/miniconda3/envs/diffusion_torch2.4/lib/python3.11/site-packages/redis/connection.py", line 477, in send_command                                                                                                                                                                                                                                    
[rank14]:     self._command_packer.pack(*args),                                                                                                                                                                                                                                                                                                                                           
[rank14]:     ^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^                                                                                                                                                                                                                                                                                                                                            
[rank14]:   File "/home/ubuntu/efs_gpu/miniconda3/envs/diffusion_torch2.4/lib/python3.11/site-packages/redis/connection.py", line 71, in pack                                                                                                                                                                                                                                             
[rank14]:     raise DataError(value).with_traceback(traceback)                                                                                                                                                                                                                                                                                                                            
[rank14]:   File "/home/ubuntu/efs_gpu/miniconda3/envs/diffusion_torch2.4/lib/python3.11/site-packages/redis/connection.py", line 68, in pack                                                                                                                                                                                                                                             
[rank14]:     output.append(hiredis.pack_command(args))                                                                                                                                                                                                                                                                                                                                   
[rank14]:                   ^^^^^^^^^^^^^^^^^^^^^^^^^^                                                                                                                                                                                                                                                                                                                                    
[rank14]: redis.exceptions.DataError: A tuple item must be str, int, float or bytes.
```

bhack commented on test/inductor/mock_cache.py:
```diff
@@ -1,127 +1,114 @@
 
-class _MockRedisRemoteCache:
-    _state = _MockCacheState("Redis")
+global_stats = _GlobalStats()
```
This is not indented correctly.

oulgen replied:
What do you mean, this is intended to be top level

bhack replied:
See https://github.com/pytorch/pytorch/actions/runs/10602712425/job/29386272600?pr=134032#step:20:645

bhack replied:
it was on line 108. But now we have new push...

aorenste replied:
You are correct. I threw in some debugging prints to try to figure out why a test is failing and messed up the indent on them. Unfortunately it's an internal PR which sometimes takes some time to push changes out.

bhack commented on test/inductor/mock_cache.py:
```diff
@@ -1,127 +1,114 @@
-        assert isinstance(key, (bytes, str))
+class MockBackend(RemoteCacheBackend[Any]):
+    def __init__(self, name: str, cache: Dict[str, object]) -> None:
+        print(f"*** MockBackend.__init__({name})",file=sys.stderr)
```
Missing space after the coma. 
There are few other of these.

aorenste replied:
This is just debug cruft to try to figure out why some tests are failing in CI.
This is still a work in progress. Please don't waste energy reviewing it.

bhack replied:
You can cherry-pick 2f44196c203d4c7b853118d0a08121d174d006e0

bhack replied:
>  This is still a work in progress. Please don't waste energy reviewing it.

Oh I supposed we used the DRAFT status of the Github PR when not ready for review. But probably with the export is different.

bhack replied:
Let me know when it is done I will take the last commit to check what is the effect on 2.4.0 branch on the CI and its compatibility.

bhack commented:
Who is going to create an instance of the `redis-server` in the CI?

aorenste commented:
> Who is going to create an instance of the `redis-server` in the CI?

The latest version shouldn't need it - I mock out the backend interaction at a lower level than we used to - but still above needing an actual redis server.

bhack commented:
But `self._redis` calls are still requiring a real server or not?

aorenste commented:
> But `self._redis` calls are still requiring a real server or not?

No self._redis calls should be made during tests - before starting up we fill in backend_override_cls so we never actually create a RedisRemoteCacheBackend. This could have been done by mocking RedisRemoteCacheBackend too - but then we lose the ability to detect *which* cache is accessing the cache (fx_cache, triton, autotune) - we would just know that the cache was accessed.

bhack commented:
Now I understand I was asking cause I didn't see the override for `RedisRemoteCacheBackend`.

bhack commented:
Is it safe enough to test it with `MockBackend` instead of the real server/client?

aorenste commented:
It depends on your definition of "safe" and what we're trying to actually get out of it.  Fundamentally a backend is responsible for putting and getting some raw data - just a fancy Dict.  If we test at too high a level then we miss some of our code (whatever is below the testing).  If we test at too low a level then we add a lot of complexity (do we need a different server for each test? Is there any problems with parallel servers?)

I think in this case we want to make sure that the backends (conceptually / API) are as simple as possible - so for most of the testing whether the backend is a full redis server or a simple Dict it doesn't really matter. Adding unit tests to make sure that a backend can successfully talk to its backend properly is a separate (but also somewhat important) task.

bhack commented:
I meant how the "simulation" is safe enough to de-facto not test `redis.set` and `redis.get` (in the REDIS case).

aorenste commented:
IMO it should be safe for this set of tests. I think testing that testing the implementation of `redis.set` and `redis.get` is more appropriate for a redis backend-specific unit test as opposed to during the testing of remote caching in general.

bhack commented:
So could we think currently to the`MockBackend` as a common (fake) intermediate representation of the remote cache across the backends and then it will require vertical integration tests?

aorenste commented:
> So could we think currently to the`MockBackend` as a common (fake) intermediate representation of the remote cache across the backends and then it will require vertical integration tests?

That's right.

bhack commented:
`binary-manywheel` failures aren't related to this PR. See https://github.com/pytorch/pytorch/issues/134236

aorenste commented:
Passes all tests - ready for review.

oulgen commented:
Looks good to me, you can go ahead and land it, already approved

aorenste committed:
Clean up RemoteCache classes (#134032)

Summary:
Pull Request resolved: https://github.com/pytorch/pytorch/pull/134032

The existing RemoteCacheBackend classes were a bit haphazard - some of them accepted bytes only, some accepted objects, some returned different types of objects than were passed in.

Update them to be more consistent:

1. RemoteCacheBackend is an implementation of a backend: Redis, Memcache, Manifold, LocalFile

2. RemoteCacheSerde is an implementation of a serde protocol - to turn structured objects (dict, list, etc) into bytes: RemoteCacheJsonSerde (json encoding), RemoteCachePassthroughSerde (strictly bytes only)

3. RemoteCache is the cache implementation itself, mixing a RemoteCacheBackend along with an RemoteCacheSerde to provide structured caching.

Other than simply reorganizing the existing cache code this also fixes the Redis autotune caching for OSS.

Test Plan: unit tests

Reviewed By: oulgen

Differential Revision: D61178859
